#631/cypress registration formular

sebastian-vass opened the pull request:
- Creating of a test for checking registration flow.
  - mobile ✅ 
  - desktop ✅ 
-  Added visual tests for this flow.
- Update the configuration, add new data-cy attributes, and prepare for running tests in CI.

sebastian-vass commented on tests/cypress.config.ts:
```diff
@@ -1,15 +1,28 @@
       config.baseUrl = config.env.BASEURL
-      config.viewportWidth = 1920;
-      config.viewportHeight = 1080;
+      config.viewportWidth = 1920
+      config.viewportHeight = 1080
```
Delete this before merge.

mpinter commented on tests/.prettierignore:
```diff
@@ -0,0 +1,6 @@
+yarn.lock
+.next
+
+backend/forms/*/form.*.sef.json
```
I guess this isn't needed ?

sebastian-vass replied:
Sorry my mistake, I was copying prittier from `/next` project so now we should have it unified across the project.

mpinter commented on next/components/forms/widget-components/InputField/InputField.tsx:
```diff
@@ -156,7 +156,7 @@ const InputField = forwardRef<HTMLInputElement, InputFieldProps>(
             </span>
           )}
-          <input {...inputProps} ref={ref} name={inputProps.id} className={style} />
+          <input {...inputProps} ref={ref} name={inputProps.id} className={style} data-cy={`input-${inputProps.name}`}/>
```
do these make sense when they are dynamic ? I understand they help you target directly with just this, put since you're relying on the non-cypress world for the naming, why not select using `data-cy` together with the value of `name` param for example ?

same question elsewhere on the form components

mpinter reviewed (changes requested):
I've rebased the changes on top of master (sorry for break of etiquette! 🙂 you'll need to reset/force-pull locally) and updated the node version which happened in the meantime. Don't know if the issues are because of that, but either way:

- I get bunch of errors when running locally (first screenshot)
- each run generates new files (not ginitignored - second screenshot) - if it's because of the first point then disregard

<img width="1072" alt="Screenshot 2023-12-01 at 08 44 57" src="https://github.com/bratislava/konto.bratislava.sk/assets/2327312/8b972a35-3629-4ac9-875f-2bb89532a9e4">
<img width="560" alt="Screenshot 2023-12-01 at 08 51 35" src="https://github.com/bratislava/konto.bratislava.sk/assets/2327312/da8ae4bd-2809-40ea-8b1e-02f30dc811aa">

sebastian-vass commented:
> I've rebased the changes on top of master (sorry for break of etiquette! 🙂 you'll need to reset/force-pull locally) and updated the node version which happened in the meantime. Don't know if the issues are because of that, but either way:
> 
> * I get bunch of errors when running locally (first screenshot)
> * each run generates new files (not ginitignored - second screenshot) - if it's because of the first point then disregard

ad 1.) Due to node upgrade to v20 and also there is probably a problem with border/color change for error inputs.
ad 2.) I put those gitignors into an empty file on purpose, I just don't understand why the project is trying to delete them.

mpinter reviewed (changes requested):
Test failing on local machine, output:

```
press-registration-formular ✓ ➭ yarn cypress:run
yarn run v1.22.19
warning package.json: No license field
$ ./cypress.sh run chrome all

DevTools listening on ws://127.0.0.1:55572/devtools/browser/86b32e46-6a6e-4c32-817d-ef87751fe6f3
2023-12-05 09:43:42.586 Cypress[30434:421957] WARNING: Secure coding is not enabled for restorable state! Enable secure coding by implementing NSApplicationDelegate.applicationSupportsSecureRestorableState: and returning YES.

====================================================================================================

  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:        13.6.0                                                                         │
  │ Browser:        Chrome 119 (headless)                                                          │
  │ Node Version:   v20.10.0 (/Users/martinpinter/.volta/tools/image/node/20.10.0/bin/node)       │
  │                 39m                                                                            │
  │ Specs:          2 found (registration/registration.cy.ts, smokeTests/pageStatusVerification.cy │
  │                 .ts)                                                                           │
  │ Searched:       cypress/e2e/**/*.cy.{js,jsx,ts,tsx}                                            │
  │ Experiments:    experimentalRunAllSpecs=true                                                   │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                    
  Running:  registration/registration.cy.ts                                                 (1 of 2)


  Registration flow
    desktop
      (Attempt 1 of 2) Submitting a empty registration form and check validation.
      1) Submitting a empty registration form and check validation.
      ✓ Filling out the registration form. (2302ms)
      (Attempt 1 of 2) Check that required inputs are not in error state.
      2) Check that required inputs are not in error state.
      ✓ Submitting the form and checking the redirection to 2FA. (105ms)
      (Attempt 1 of 2) Check the 2FA page.
      3) Check the 2FA page.
    mobile
      (Attempt 1 of 2) Submitting a empty registration form and check validation.
      ✓ Submitting a empty registration form and check validation. (3829ms)
      ✓ Filling out the registration form. (2099ms)
      (Attempt 1 of 2) Check that required inputs are not in error state.
      4) Check that required inputs are not in error state.
      ✓ Submitting the form and checking the redirection to 2FA. (91ms)
      (Attempt 1 of 2) Check the 2FA page.
      5) Check the 2FA page.


  5 passing (50s)
  5 failing

  1) Registration flow
       desktop
         Submitting a empty registration form and check validation.:
     Error: Image diff factor (0.124%) is bigger than maximum threshold option 0.01.
Warning: Images size mismatch - new screenshot is 696px by 1181px while old one is 696px by 1116 (width x height).
      at Context.eval (webpack://cypress/./node_modules/@frsource/cypress-plugin-visual-regression-diff/dist/support.js:165:0)

  2) Registration flow
       desktop
         Check that required inputs are not in error state.:
     Error: Image diff factor (0.152%) is bigger than maximum threshold option 0.01.
Warning: Images size mismatch - new screenshot is 696px by 1125px while old one is 696px by 1004 (width x height).
      at Context.eval (webpack://cypress/./node_modules/@frsource/cypress-plugin-visual-regression-diff/dist/support.js:165:0)

  3) Registration flow
       desktop
         Check the 2FA page.:
     AssertionError: Timed out retrying after 4000ms: Expected to find element: `[data-cy=verification-description]`, but never found it.
      at Context.eval (webpack://cypress/./cypress/support/commands.ts:6:0)

  4) Registration flow
       mobile
         Check that required inputs are not in error state.:
     Error: Image diff factor (0.155%) is bigger than maximum threshold option 0.01.
Warning: Images size mismatch - new screenshot is 360px by 1084px while old one is 360px by 1008 (width x height).
      at Context.eval (webpack://cypress/./node_modules/@frsource/cypress-plugin-visual-regression-diff/dist/support.js:165:0)

  5) Registration flow
       mobile
         Check the 2FA page.:
     AssertionError: Timed out retrying after 4000ms: Expected to find element: `[data-cy=verification-description]`, but never found it.
      at Context.eval (webpack://cypress/./cypress/support/commands.ts:6:0)




  (Results)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Tests:        10                                                                               │
  │ Passing:      5                                                                                │
  │ Failing:      5                                                                                │
  │ Pending:      0                                                                                │
  │ Skipped:      0                                                                                │
  │ Screenshots:  17                                                                               │
  │ Video:        false                                                                            │
  │ Duration:     49 seconds                                                                       │
  │ Spec Ran:     registration/registration.cy.ts                                                  │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


  (Screenshots)

  -  /Users/martinpinter/code/konto.bratislava.sk/tests/cypress/e2e/registration/__im     (696x1004)
     age_snapshots__/Registration flow desktop Submitting a empty registration form a               
     nd check validation #0.actual.png                                                              
  -  /Users/martinpinter/code/konto.bratislava.sk/tests/cypress/e2e/registration/__im     (696x1181)
     age_snapshots__/Registration flow desktop Submitting a empty registration form a               
     nd check validation #1.actual.png                                                              
  -  /Users/martinpinter/code/konto.bratislava.sk/tests/cypress/screenshots/registrat     (1280x591)
     ion/registration.cy.ts/Registration flow -- desktop -- Submitting a empty regist               
     ration form and check validation (failed).png                                                  
  -  /Users/martinpinter/code/konto.bratislava.sk/tests/cypress/screenshots/registrat     (1280x591)
     ion/registration.cy.ts/Registration flow -- desktop -- Submitting a empty regist               
     ration form and check validation (failed) (attempt 2).png                                      
  -  /Users/martinpinter/code/konto.bratislava.sk/tests/cypress/e2e/registration/__im     (696x1125)
     age_snapshots__/Registration flow desktop Check that required inputs are not in                
     error state #0.actual.png                                                                      
  -  /Users/martinpinter/code/konto.bratislava.sk/tests/cypress/screenshots/registrat     (1280x591)
     ion/registration.cy.ts/Registration flow -- desktop -- Check that required input               
     s are not in error state (failed).png                                                          
  -  /Users/martinpinter/code/konto.bratislava.sk/tests/cypress/screenshots/registrat     (1280x591)
     ion/registration.cy.ts/Registration flow -- desktop -- Check that required input               
     s are not in error state (failed) (attempt 2).png                                              
  -  /Users/martinpinter/code/konto.bratislava.sk/tests/cypress/screenshots/registrat     (1280x591)
     ion/registration.cy.ts/Registration flow -- desktop -- Check the 2FA page (faile               
     d).png                                                                                         
  -  /Users/martinpinter/code/konto.bratislava.sk/tests/cypress/screenshots/registrat     (1280x591)
     ion/registration.cy.ts/Registration flow -- desktop -- Check the 2FA page (faile               
     d) (attempt 2).png                                                                             
  -  /Users/martinpinter/code/konto.bratislava.sk/tests/cypress/e2e/registration/__im     (360x1008)
     age_snapshots__/Registration flow mobile Submitting a empty registration form an               
     d check validation #0.actual.png                                                               
  -  /Users/martinpinter/code/konto.bratislava.sk/tests/cypress/e2e/registration/__im     (360x1104)
     age_snapshots__/Registration flow mobile Submitting a empty registration form an               
     d check validation #1.actual.png                                                               
  -  /Users/martinpinter/code/konto.bratislava.sk/tests/cypress/screenshots/registrat     (1280x591)
     ion/registration.cy.ts/Registration flow -- mobile -- Submitting a empty registr               
     ation form and check validation (failed).png                                                   
  -  /Users/martinpinter/code/konto.bratislava.sk/tests/cypress/e2e/registration/__im     (360x1084)
     age_snapshots__/Registration flow mobile Check that required inputs are not in e               
     rror state #0.actual.png                                                                       
  -  /Users/martinpinter/code/konto.bratislava.sk/tests/cypress/screenshots/registrat     (1280x591)
     ion/registration.cy.ts/Registration flow -- mobile -- Check that required inputs               
      are not in error state (failed).png                                                           
  -  /Users/martinpinter/code/konto.bratislava.sk/tests/cypress/screenshots/registrat     (1280x591)
     ion/registration.cy.ts/Registration flow -- mobile -- Check that required inputs               
      are not in error state (failed) (attempt 2).png                                               
  -  /Users/martinpinter/code/konto.bratislava.sk/tests/cypress/screenshots/registrat     (1280x591)
     ion/registration.cy.ts/Registration flow -- mobile -- Check the 2FA page (failed               
     ).png                                                                                          
  -  /Users/martinpinter/code/konto.bratislava.sk/tests/cypress/screenshots/registrat     (1280x591)
     ion/registration.cy.ts/Registration flow -- mobile -- Check the 2FA page (failed               
     ) (attempt 2).png                                                                              


────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                    
  Running:  smokeTests/pageStatusVerification.cy.ts                                         (2 of 2)


  Page status verification
    ✓ Verification of successful page loading (1501ms)


  1 passing (2s)


  (Results)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Tests:        1                                                                                │
  │ Passing:      1                                                                                │
  │ Failing:      0                                                                                │
  │ Pending:      0                                                                                │
  │ Skipped:      0                                                                                │
  │ Screenshots:  0                                                                                │
  │ Video:        false                                                                            │
  │ Duration:     1 second                                                                         │
  │ Spec Ran:     smokeTests/pageStatusVerification.cy.ts                                          │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


====================================================================================================

  (Run Finished)


       Spec                                              Tests  Passing  Failing  Pending  Skipped  
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ ✖  registration/registration.cy.ts          00:49       10        5        5        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  smokeTests/pageStatusVerification.c      00:01        1        1        -        -        - │
  │    y.ts                                                                                        │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
    ✖  1 of 2 failed (50%)                      00:51       11        6        5        -        -  

error Command failed with exit code 5.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
